cupyx/distributed/array/_array.py

```diff
@@ -0,0 +1,917 @@
+        raise NotImplementedError(
+            'DistributedArray currently does not support device.')
+
+    @property
+    def flags(self):
+        """Not supported."""
+        raise NotImplementedError(
+            'DistributedArray currently does not support flags.')
+
+    @property
+    def flat(self):
+        """Not supported."""
+        raise NotImplementedError(
+            'DistributedArray currently does not support flat.')
+
+    @property
+    def imag(self):
+        """Not supported."""
+        raise NotImplementedError(
+            'DistributedArray currently does not support imag.')
+
+    @property
+    def real(self):
+        """Not supported."""
+        raise NotImplementedError(
+            'DistributedArray currently does not support real.')
+
+    @property
+    def shape(self):
+        """Tuple of array dimensions.
+
+        Assignment to this property is currently not supported.
+
+        .. seealso: :attr:`cupy.ndarray.shape`, :attr:`numpy.ndarray.shape`
+
+        """
+        return super().shape
+
+    @shape.setter
+    def shape(self, newshape):
+        raise NotImplementedError(
+            'DistributedArray currently does not support assignment to shape.')
+
+    @property
+    def strides(self):
+        """Not supported."""
+        raise NotImplementedError(
+            'DistributedArray currently does not support strides.')
+
+
+def distributed_array(
+    array: ArrayLike,
+    index_map: dict[int, Any],
+    mode: str = 'replica',
+) -> DistributedArray:
+    """Creates a distributed array from the given data.
+
+    This function does not check if all elements of the given array are stored
+    in some of the chunks.
+
+    Args:
+        array (array_like): :class:`DistributedArray` object,
+            :class:`cupy.ndarray` object or any other object that can be passed
+            to :func:`numpy.array`.
+        index_map (dict from int to array indices): Indices for the chunks
+            that devices with designated IDs own. One device can have multiple
+            chunks, which can be specified as a list of array indices.
+        mode ({'replica', 'min', 'max', 'sum', 'prod'}): How overlaps of the
+            chunks are interpreted.
+
+    .. seealso::
+            :attr:`DistributedArray.mode` for details about modes.
+
+    Example:
+        >>> array = cupy.arange(9).reshape(3, 3)
+        >>> A = distributed_array(
+        ...     array,
+        ...     {0: [(slice(2), slice(2)),  # array[:2, :2]
+        ...          slice(None, None, 2)], # array[::2]
+        ...      1:  (slice(1, None), 2)})  # array[1:, 2]
+    """
+    if isinstance(array, DistributedArray):
+        if array.mode != mode:
+            array = array.change_mode(mode)
+        if array.index_map != index_map:
+            array = array.reshard(index_map)
+        return DistributedArray(
+            array.shape, array.dtype, array._chunks_map, array._mode,
+            array._comms)
+
```

Comment:
TODO
Support regular CuPy arrays and distribute the chunks without needing to synchronize